client/fingerprint/env_gce.go

```diff
@@ -0,0 +1,236 @@
+package fingerprint
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/nomad/client/config"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+type GCEMetadataClient struct {
+	*http.Client
+	logger      *log.Logger
+	metadataURL string
+}
+
+type ReqError struct {
+	StatusCode int
+}
+
+func (e ReqError) Error() string {
+	return http.StatusText(e.StatusCode)
+}
+
+func NewGCEMetadataClient(logger *log.Logger) GCEMetadataClient {
+	// Read the internal metadata URL from the environment, allowing test files to
+	// provide their own
+	metadataURL := os.Getenv("GCE_ENV_URL")
+	if metadataURL == "" {
+		metadataURL = "http://169.254.169.254/computeMetadata/v1/instance/"
+	}
+
+	// assume 2 seconds is enough time for inside GCE network
+	client := &http.Client{
+		Timeout: 2 * time.Second,
+	}
+
+	return GCEMetadataClient{
+		Client:      client,
+		logger:      logger,
+		metadataURL: metadataURL,
+	}
+}
+
+func (g GCEMetadataClient) Get(attribute string, recursive bool) (string, error) {
+	reqUrl := g.metadataURL + attribute
+	if recursive {
+		reqUrl = reqUrl + "?recursive=true"
+	}
+
+	parsedUrl, err := url.Parse(reqUrl)
+	if err != nil {
+		return "", err
+	}
+
+	req := &http.Request{
+		Method: "GET",
+		URL:    parsedUrl,
+		Header: http.Header{
+			"Metadata-Flavor": []string{"Google"},
+		},
+	}
+
+	res, err := g.Client.Do(req)
+	if err != nil {
+		return "", err
+	}
+
+	resp, err := ioutil.ReadAll(res.Body)
+	res.Body.Close()
+	if err != nil {
+		g.logger.Printf("[ERR]: fingerprint.env_gce: Error reading response body for GCE %s", attribute)
+		return "", err
+	}
+
+	if res.StatusCode >= 400 {
+		return "", ReqError{res.StatusCode}
+	}
+
+	return string(resp), nil
+}
+
+// EnvGCEFingerprint is used to fingerprint the CPU
+type EnvGCEFingerprint struct {
+	logger *log.Logger
+}
+
+// NewEnvGCEFingerprint is used to create a CPU fingerprint
+func NewEnvGCEFingerprint(logger *log.Logger) Fingerprint {
+	f := &EnvGCEFingerprint{logger: logger}
+	return f
+}
+
+func checkError(err error, logger *log.Logger, desc string) error {
+	// If it's a URL error, assume we're not actually in an GCE environment.
+	// To the outer layers, this isn't an error so return nil.
+	if _, ok := err.(*url.Error); ok {
+		logger.Printf("[ERR] fingerprint.env_gce: Error querying GCE " + desc + ", skipping")
+		return nil
+	}
+	// Otherwise pass the error through.
+	return err
+}
+
+func (f *EnvGCEFingerprint) Fingerprint(cfg *config.Config, node *structs.Node) (bool, error) {
+	if !f.isGCE() {
+		return false, nil
+	}
+
+	// newNetwork is populated and addded to the Nodes resources
+	newNetwork := &structs.NetworkResource{
+		Device: "eth0",
+	}
+
+	if node.Links == nil {
+		node.Links = make(map[string]string)
+	}
+
+	client := NewGCEMetadataClient(f.logger)
```

Comment:
Good idea. I originally did it that way so that the client would stay around forever once it was n longer needed, but I see that the Fingerprint objects aren't persistent anyway.